torch/optim/adam.py

```diff
@@ -557,19 +587,22 @@ def _fused_adam(
 ) -> None:
     if not params:
         return
+    if differentiable:
+        raise RuntimeError("_fused_adam is not differentiable")
     grad_scale_dict = {grad_scale.device: grad_scale} if grad_scale is not None else None
     found_inf_dict = {found_inf.device: found_inf} if found_inf is not None else None
     grouped_tensors = Optimizer._group_tensors_by_device_and_dtype(
         [params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps])
-    for (device, dtype) in grouped_tensors:
-        ((
-            device_params,
-            device_grads,
-            device_exp_avgs,
-            device_exp_avg_sqs,
-            device_max_exp_avg_sqs,
-            device_state_steps,
-        ), _) = grouped_tensors[(device, dtype)]
+    for (device, _), ((device_params,
+                       device_grads,
+                       device_exp_avgs,
+                       device_exp_avg_sqs,
+                       device_max_exp_avg_sqs,
+                       device_state_steps,), _) in grouped_tensors.items():
+        if not torch.is_tensor(lr):
+            lr = torch.tensor(lr, device=device, dtype=torch.float)
+        elif lr.device != device:
+            lr = lr.to(device=device)
```

Comment:
Cross device?? Most likely no no?

Reply:
Why is this bad? Don't we do something similar for grad_scale and found_inf?

Reply:
Revised impl to not need this by adding an overload